### PR TITLE
raft: Avoid holding unneeded memory in unstable log's entries array

### LINKED ISSUE #7746: raft: memory "leak" in unstable log
`unstable.stableTo` truncates a prefix of the unstable log after those entries have been written to stable storage:

```
func (u *unstable) stableTo(i, t uint64) {
	gt, ok := u.maybeTerm(i)
	if !ok {
		return
	}
	// if i < offset, term is matched with the snapshot
	// only update the unstable entries if term is matched with
	// an unstable entry.
	if gt == t && i >= u.offset {
		u.entries = u.entries[i+1-u.offset:]
		u.offset = i + 1
	}
}
```

The truncation is performed by the line: `u.entries = u.entries[i+1-u.offset:]`. But consider what happens to the truncated entries (i.e. `u.entries[:i+1-u.offset]`). They are left untouched and continue to reference `raftpb.Entry.Data` preventing GC from reclaiming that memory. 

On the surface it seems straightforward to clear those entries via something like:

```diff
diff --git a/raft/log_unstable.go b/raft/log_unstable.go
index 8ae301c..a64349c 100644
--- a/raft/log_unstable.go
+++ b/raft/log_unstable.go
@@ -83,7 +83,14 @@ func (u *unstable) stableTo(i, t uint64) {
        // only update the unstable entries if term is matched with
        // an unstable entry.
        if gt == t && i >= u.offset {
-               u.entries = u.entries[i+1-u.offset:]
+               // Clear the entries we are truncating so that the GC can reclaim any
+               // resources they were pointing to. This is important if one or more of the
+               // truncated entries is large.
+               n := i + 1 - u.offset
+               for j := uint64(0); j < n; j++ {
+                       u.entries[j] = pb.Entry{}
+               }
+               u.entries = u.entries[n:]
                u.offset = i + 1
        }
 }
```

Unfortunately, this causes `TestUnstableEnts` to fail as we're mutating the entries in `unstable` which is one of the guarantees that test is checking is not violated. I'm not clear on whether that is accidental or not.

### PR DESCRIPTION
**Pull request by a-robinson**
Accumulation of old entries in the underlying array backing the
entries slice has been found to cause massive memory growth in
CockroachDB for workloads that do large (1MB) writes
(https://github.com/cockroachdb/cockroach/issues/14776)

This doesn't appear to have much consistent effect on the raft
benchmarks, although it's worth noting that they vary quite a bit
between runs so it's kind of tough to draw strong conclusions from them.
Let me know if there are any different benchmarks you'd like me to run!

```
benchmark              old ns/op     new ns/op     delta
BenchmarkOneNode-8     3283          3125          -4.81%

benchmark              old allocs     new allocs     delta
BenchmarkOneNode-8     6              6              +0.00%

benchmark              old bytes     new bytes     delta
BenchmarkOneNode-8     796           727           -8.67%

benchmark                     old ns/op     new ns/op     delta
BenchmarkProposal3Nodes-8     4269          4337          +1.59%

benchmark                     old allocs     new allocs     delta
BenchmarkProposal3Nodes-8     15             13             -13.33%

benchmark                     old bytes     new bytes     delta
BenchmarkProposal3Nodes-8     5839          4544          -22.18%
```

Fixes #7746

### CODE REVIEW

**Comment by a-robinson** [comment]
@xiang90

File: raft/log_unstable.go [hunk 1]
```diff
@@ -85,6 +85,17 @@ func (u *unstable) stableTo(i, t uint64) {
 	if gt == t && i >= u.offset {
 		u.entries = u.entries[i+1-u.offset:]
 		u.offset = i + 1
+		// Discard the underlying array used by the slice if most of it isn't being
```
Comment (xiang90) [inline raft/log_unstable.go]: add a func: u.shrinkEntriesSlice?

File: raft/log_unstable.go [hunk 2]
```diff
@@ -85,6 +85,17 @@ func (u *unstable) stableTo(i, t uint64) {
...
+		// used to avoid holding onto a bunch of potentially large entries that
+		// aren't needed anymore. Simply clearing the entries wouldn't be safe
+		// because clients might still be using them.
+		if len(u.entries) == 0 {
+			u.entries = nil
+		} else if len(u.entries) < cap(u.entries)/4 {
```
Comment (xiang90) [inline raft/log_unstable.go]: why do we choose cap/4? not cap/2 or cap/8 or other value?

**Comment by xiang90** [comment]
looks good in general.

File: raft/log_unstable.go (hunk 1 above)
Comment (a-robinson) [inline raft/log_unstable.go]: Done

File: raft/log_unstable.go (hunk 2 above)
Comment (a-robinson) [inline raft/log_unstable.go]: It's pretty arbitrary. I've added a comment explaining the choice, but having a good performance benchmark to justify a constant here would be nice.

**Comment by xiang90** [comment]
lgtm

**Comment by a-robinson** [comment]
cc @bdarnell @petermattis for thoughts on the choice of constant

**Comment by bdarnell** [comment]
`cap/2` seems like the most natural choice to me, to ensure that we have no more memory in "old" entries than new ones (and it's similar to the common heuristic for arrays to grow by doubling). I think we'd only want to use a higher divisor if there are benchmarks showing the reduced allocations matter (which seems unlikely, since we discard the entire backing array whenever it reaches zero, forcing an allocation on the next append).

**Comment by petermattis** [comment]
`cap/2` sounds good to me as well without other evidence that a different divisor is better.

**Comment by a-robinson** [comment]
Sure, we should probably stick with the default of 2 unless we have proof something else is better.

**Comment by a-robinson** [comment]
Alright, this is ready for merge on our end then :)

**Pull request merged**